### PR TITLE
IncrementalCompact: fix stateful behavior when using multiple iterators.

### LINKED ISSUE #46945: IncrementalCompact re-iteration is broken
Normally it should be possible to (partially) iterate an object, and iterate anew using a different iterator. This is not the case with IncrementalCompact.

Set-up, on latest master:

```julia
julia> foo(i) = i == 1 ? 1 : 2
foo (generic function with 1 method)

julia> ir = only(Base.code_ircode(foo, (Int,)))[1]
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
  2 ─      return 1                                                                         │
  3 ─      return 2                                                                         │


julia> compact = Core.Compiler.IncrementalCompact(ir)
──────────────────────────────────────────────────────────────────────────────────────────────────
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
  2 ─      return 1                                                                         │
  3 ─      return 2                                                                         │
```

Let's iterate a bit:

```julia
julia> x = Core.Compiler.iterate(compact)
(Pair{Pair{Int64, Int64}, Any}(1 => 1, :(_2 === 1)), (2, 1))

julia> compact
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
──────────────────────────────────────────────────────────────────────────────────────────────────
  └──      goto #3 if not %1                                                                │
  2 ─      return 1                                                                         │
  3 ─      return 2                                                                         │


julia> x = Core.Compiler.iterate(compact, x[2])
(Pair{Pair{Int64, Int64}, Any}(2 => 2, :(goto %3 if not %1)), (3, 2))

julia> compact
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
──────────────────────────────────────────────────────────────────────────────────────────────────
  2 ─      return 1                                                                         │
  3 ─      return 2                                                                         │
```

The same, but with a new iterator:

```julia
julia> x = Core.Compiler.iterate(compact)
(Pair{Pair{Int64, Int64}, Any}(3 => 3, :(return 1)), (4, 1))

julia> compact
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
  2 ─      return 1                                                                         │
──────────────────────────────────────────────────────────────────────────────────────────────────
  3 ─      return 2                                                                         │


julia> x = Core.Compiler.iterate(compact, x[2])
(Pair{Pair{Int64, Int64}, Any}(4 => 4, :(return 2)), (5, 1))

julia> x = Core.Compiler.iterate(compact, x[2])

julia> compact
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
  2 ─      return 1                                                                         │
  3 ─      return 2                                                                         │
──────────────────────────────────────────────────────────────────────────────────────────────────
```

So this second iteration resumed where the previous iterator left off, which is wrong, but it also breaks the resulting IR's CFG:

```julia
julia> Core.Compiler.complete(compact)
1 1 ─ %1 = (_2 === 1)::Bool                                                                 │╻ ==
  └──      goto #3 if not %1                                                                │
  !!! ─      return 1                                                                       │
  !!! ─      return 2                                                                       │
```

The `!!!`s don't occur if I only iterate using a single iterator.

### PR DESCRIPTION
**Pull request by maleadt**
Fixes https://github.com/JuliaLang/julia/issues/46945

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by Keno** [comment]
Seems good to me. I think there's some argument that maybe it should restart at the beginning and iterate the already-compacted stmts first, but I'm ok with the stateful version also.

**Comment by maleadt** [comment]
This crashes during bootstrap, because in inlining we have some special IncrementalCompact usage sharing state: https://github.com/JuliaLang/julia/blob/9fd408723aab0974f399b8ffc952af4b4d20b6f7/base/compiler/ssair/inlining.jl#L422-L423
https://github.com/JuliaLang/julia/blob/9fd408723aab0974f399b8ffc952af4b4d20b6f7/base/compiler/ssair/ir.jl#L630-L643

**Comment by maleadt** [comment]
Found the issue: the iterator state was tracking the active bb, not the active result bb, so I added a field to IncrementalCompact to keep track of that, and use it during initialization of the iterator.

Now, with `active_bb` being part of `IncrementalCompact`, we can remove the actual iterator state completely. Since that code is pretty hot, let's see how it affects performance:

@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46952#issuecomment-1268272767) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3b2d5d0_vs_f056dbc/report.md).

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -1420,18 +1421,23 @@ function iterate(it::CompactPeekIterator, (idx, aidx, bidx)::NTuple{3, Int}=(it.
     return (compact.ir.stmts[idx][:inst], (idx + 1, aidx, bidx))
 end
 
-# This Union{Nothing, Pair{Pair{Int,Int},Any}} cannot be stack allocated, so we inline it
-@inline function iterate(compact::IncrementalCompact, st::Tuple{Int, Int}=(compact.idx, 1))
-    st = iterate_compact(compact, st)
+struct IncrementalCompactState end
```
Comment (KristofferC) [inline base/compiler/ssair/ir.jl:1424]: Is it worth being this explicit over e.g. just using `nothing` as the state?

Comment (maleadt) [reply base/compiler/ssair/ir.jl]: Ah yeah, of course, I guess I was confused by `iterate(...) = nothing` being indicative of the end-of-iteration, but here we return `something, nothing`.

**Comment by maleadt** [comment]
Let's see if this reproduces if we test against the exact merge-base:

@nanosoldier `runbenchmarks("inference", vs="@f056dbc78172eb1aec1a3b41a4f9b15d3683306e")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46952#issuecomment-1268895405) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/7164849_vs_f056dbc/report.md).

**Comment by gbaraldi** [comment]
Seems to be better or equal and fixes the issue so SGTM

**Pull request merged**